Index: lib/Target/ARM/MCTargetDesc/ARMMCAsmInfo.cpp
===================================================================
--- lib/Target/ARM/MCTargetDesc/ARMMCAsmInfo.cpp	(revision 142659)
+++ lib/Target/ARM/MCTargetDesc/ARMMCAsmInfo.cpp	(working copy)
@@ -77,6 +77,8 @@
   HasLEB128 = true;
   SupportsDebugInformation = true;
 
+  StructorOutputOrder = Structors::PriorityOrder;
+
   // Exceptions handling
   if (EnableARMEHABI)
     ExceptionsType = ExceptionHandling::ARM;
