Index: tools/clang/lib/Driver/Driver.cpp
===================================================================
--- tools/clang/lib/Driver/Driver.cpp	(revision 142659)
+++ tools/clang/lib/Driver/Driver.cpp	(working copy)
@@ -1571,7 +1571,11 @@
   case llvm::Triple::Minix:
     return createMinixHostInfo(*this, Triple);
   case llvm::Triple::Linux:
-    return createLinuxHostInfo(*this, Triple);
+    // HACK: I could not convince LinuxHostInfo to work with a GCC cross-toolchain.
+    if (Triple.getArch() == llvm::Triple::arm)
+      return createUnknownHostInfo(*this, Triple);
+    else
+      return createLinuxHostInfo(*this, Triple);
   case llvm::Triple::Win32:
     return createWindowsHostInfo(*this, Triple);
   case llvm::Triple::MinGW32:
