Index: src/c++11/system_error.cc
===================================================================
--- src/c++11/system_error.cc	(revision 197031)
+++ src/c++11/system_error.cc	(working copy)
@@ -45,6 +45,8 @@
       // _GLIBCXX_HAVE_STRERROR_L, strerror_l(i, cloc)
       return string(strerror(i));
     }
+
+    generic_error_category() {}
   };
 
   struct system_error_category : public std::error_category
@@ -60,6 +62,8 @@
       // _GLIBCXX_HAVE_STRERROR_L, strerror_l(i, cloc)
       return string(strerror(i));
     }
+
+    system_error_category() {}
   };
 
   const generic_error_category generic_category_instance;
Index: src/c++11/future.cc
===================================================================
--- src/c++11/future.cc	(revision 197031)
+++ src/c++11/future.cc	(working copy)
@@ -55,6 +55,8 @@
       }
       return __msg;
     }
+
+    future_error_category() {}
   };
 
   const future_error_category&
Index: src/c++11/random.cc
===================================================================
--- src/c++11/random.cc	(revision 197031)
+++ src/c++11/random.cc	(working copy)
@@ -52,6 +52,8 @@
       return __ret;
     }
 
+#undef _GLIBCXX_X86_RDRAND
+
 #if (defined __i386__ || defined __x86_64__) && defined _GLIBCXX_X86_RDRAND
     unsigned int
     __attribute__ ((target("rdrnd")))
