--- projects/CMakeLists.txt     (revision 160622)
+++ projects/CMakeLists.txt     (working copy)
@@ -4,8 +4,7 @@
 file(GLOB entries *)
 foreach(entry ${entries})
   if(IS_DIRECTORY ${entry} AND EXISTS ${entry}/CMakeLists.txt)
-    if((NOT ${entry} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR}/compiler-rt) AND
-       (NOT ${entry} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR}/libcxx))
+    if(NOT ${entry} STREQUAL ${CMAKE_CURRENT_SOURCE_DIR}/libcxx)
       add_subdirectory(${entry})
     endif()
   endif()
