diff -Naur gtk+-2.24.8.orig/gtk/gtkmenushell.h gtk+-2.24.8/gtk/gtkmenushell.h
--- gtk+-2.24.8.orig/gtk/gtkmenushell.h	2011-11-10 18:38:03.000000000 +0100
+++ gtk+-2.24.8/gtk/gtkmenushell.h	2012-01-12 21:13:12.851805399 +0100
@@ -135,6 +135,8 @@
 void     _gtk_menu_shell_set_keyboard_mode (GtkMenuShell *menu_shell,
                                             gboolean      keyboard_mode);
 gboolean _gtk_menu_shell_get_keyboard_mode (GtkMenuShell *menu_shell);
+gboolean ubuntu_gtk_menu_shell_activate_mnemonic (GtkMenuShell *shell, GtkWidget *item);
+gboolean ubuntu_gtk_menu_shell_activate_first (GtkMenuShell *menu_shell, gboolean search_sensitive);
 
 G_END_DECLS
 
